Improve performance on multiple versions

klauspost opened the pull request:
## Description

Existing:

``` Go
type xlMetaV2 struct {
    Versions []xlMetaV2Version `json:"Versions" msg:"Versions"`
}
```

Serialized as regular MessagePack.

Change to add a shallow header and each version separately:

```Go
//msgp:tuple xlMetaV2VersionHeader
type xlMetaV2VersionHeader struct {
	VersionID [16]byte
	ModTime   int64
	Signature [4]byte
	Type      VersionType
	Flags     xlFlags
}
```

Serialize as streaming MessagePack, format:

```
int(headerVersion)
int(xlmetaVersion)
int(nVersions)
for each version {
    binary blob, xlMetaV2VersionHeader, serialized
    binary blob, xlMetaV2Version, serialized.
}
```

xlMetaV2VersionHeader is <= 30 bytes serialized. Deserialized struct can easily be reused and does not contain pointers, so efficient as a slice (single alloc)

This allows quickly parsing everything as slices of bytes (no copy).

Versions are always *saved* sorted by modtime, newest *first*. No more need to sort on load.

* Allows checking if a version exists.
* Allows reading single version without unmarshal all.
* Allows reading latest version of type without unmarshal all.
* Allows reading latest version without unmarshal of all.
* Allows checking if latest is deletemarker by reading first entry.
* Allows adding/updating/deleting a version with only header deserialization.
* Reduces allocations on conversion to FileInfo(s).

Comparing `warp versioned --obj.size=1K --concurrent=32`

```
-------------------
Operation: DELETE
Operations: 27495 -> 36629
* Average: +33.21% (+30.4) obj/s
-------------------
Operation: GET
Operations: 123710 -> 164845
* Average: +33.25% (+0.1 MiB/s) throughput, +33.25% (+137.1) obj/s
* First Byte: Average: -3.951011ms (-17%), Median: -757.1µs (-27%), Best: +1.8µs (+0%), Worst: -506.2678ms (-16%)
-------------------
Operation: PUT
Operations: 41237 -> 54949
* Average: +33.26% (+0.0 MiB/s) throughput, +33.26% (+45.7) obj/s
-------------------
Operation: STAT
Operations: 82476 -> 109901
* Average: +33.26% (+91.4) obj/s
```

Other benchmarks below.

## How to test this PR?

Tiering and Replication should probably be checked.

## Types of changes
- [x] Optimization (provides speedup with no functional changes)

## Checklist:
- [x] Documentation updated (xl-meta executable)
- [x] Unit tests added/updated
- [x] Optimize loading existing
- [x] Optimize saving size.
- [x] Benchmarks.

klauspost committed:
Improve performance on multiple versions

Existing:

``` Go
type xlMetaV2 struct {
    Versions []xlMetaV2Version `json:"Versions" msg:"Versions"`
}
```

Serialized as regular MessagePack.

Change to add a shallow header and each version separately:

```Go
//msgp:tuple xlMetaV2VersionHeader
type xlMetaV2VersionHeader struct {
	VersionID [16]byte
	ModTime   int64
	Type      VersionType
	Flags     xlFlags
}
```

Serialize as streaming MessagePack, format:

```
int(headerVersion)
int(xlmetaVersion)
int(nVersions)
for each version {
    binary blob, xlMetaV2VersionHeader, serialized
    binary blob, xlMetaV2Version, serialized.
}
```

xlMetaV2VersionHeader is <= 30 bytes serialized. Deserialized struct can easily be reused and does not contain pointers, so efficient as a slice (single alloc)

This allows quickly parsing everything as slices of bytes (no copy).

Versions are always *saved* sorted by modtime, newest *first*. No more need to sort on load.

* Allows checking if a version exists.
* Allows reading single version without unmarshal all.
* Allows reading latest version of type without unmarshal all.
* Allows reading latest version without unmarshal of all.
* Allows checking if latest is deletemarker by reading first entry.
* Allows adding/updating/deleting a version with only header deserialization.
* Reduces allocations on conversion to FileInfo(s).

klauspost committed:
Improve performance on multiple versions

Existing:

``` Go
type xlMetaV2 struct {
    Versions []xlMetaV2Version `json:"Versions" msg:"Versions"`
}
```

Serialized as regular MessagePack.

Change to add a shallow header and each version separately:

```Go
//msgp:tuple xlMetaV2VersionHeader
type xlMetaV2VersionHeader struct {
	VersionID [16]byte
	ModTime   int64
	Type      VersionType
	Flags     xlFlags
}
```

Serialize as streaming MessagePack, format:

```
int(headerVersion)
int(xlmetaVersion)
int(nVersions)
for each version {
    binary blob, xlMetaV2VersionHeader, serialized
    binary blob, xlMetaV2Version, serialized.
}
```

xlMetaV2VersionHeader is <= 30 bytes serialized. Deserialized struct can easily be reused and does not contain pointers, so efficient as a slice (single alloc)

This allows quickly parsing everything as slices of bytes (no copy).

Versions are always *saved* sorted by modtime, newest *first*. No more need to sort on load.

* Allows checking if a version exists.
* Allows reading single version without unmarshal all.
* Allows reading latest version of type without unmarshal all.
* Allows reading latest version without unmarshal of all.
* Allows checking if latest is deletemarker by reading first entry.
* Allows adding/updating/deleting a version with only header deserialization.
* Reduces allocations on conversion to FileInfo(s).

klauspost committed:
Fix lint errors

klauspost committed:
Fix lint errors

klauspost committed:
Upgrade msgp for tests.

klauspost committed:
Always install msgp

harshavardhana commented on Makefile:
```diff
@@ -20,7 +20,7 @@ help: ## print this help
 	@mkdir -p ${GOPATH}/bin
 	@echo "Installing golangci-lint" && curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b $(GOPATH)/bin v1.40.1
-	@which msgp 1>/dev/null || (echo "Installing msgp" && go install -v github.com/tinylib/msgp@v1.1.6)
+	@echo "Installing msgp" && go install -v github.com/tinylib/msgp@v1.1.6
```
Yeah this is fine @klauspost

harshavardhana reviewed (changes requested):
Will do some testing on this PR as well.

klauspost committed:
Install specific version

klauspost committed:
Add benchmark

klauspost committed:
Add benchmark

klauspost committed:
Don't lowercase when equalfold

klauspost committed:
Don't lowercase when equalfold

klauspost committed:
Reduce a few casual allocs.

klauspost committed:
Reduce a few casual allocs.

klauspost committed:
Alloc for adding one version.
Reuse xlMetaDataDirDecoder.
Remove alloc from strings.ToLower

klauspost committed:
Alloc for adding one version.
Reuse xlMetaDataDirDecoder.
Remove alloc from strings.ToLower

klauspost committed:
Omit part etags if all empty.

klauspost committed:
Omit part etags if all empty.

klauspost committed:
Speed up legacy loading.
Remove unused code.
Add tests.

klauspost committed:
Speed up legacy loading.
Remove unused code.
Add tests.

klauspost committed:
xlMetaV2Shallow -> xlMetaV2

klauspost committed:
xlMetaV2Shallow -> xlMetaV2

klauspost committed:
Merge branch 'master' into fast-multiversion-metadata

klauspost committed:
Add missing test file.

klauspost committed:
Add missing test file.

harshavardhana committed:
Merge branch 'master' into fast-multiversion-metadata

harshavardhana commented on cmd/bucket-replication.go:
```diff
@@ -724,7 +724,7 @@ const (
 func equals(k1 string, keys ...string) bool {
 	for _, k2 := range keys {
-		if strings.EqualFold(strings.ToLower(k1), strings.ToLower(k2)) {
+		if strings.EqualFold(k1, k2) {
```
isn't it faster to just do 

```
if strings.ToLower(k1) == strings.ToLower(k2) {
}
```

klauspost replied:
@harshavardhana No, `ToLower` allocates.

klauspost replied:
(also EqualFold can abort as soon as a mismatch is found, not needing to lowercase the rest)

harshavardhana replied:
okay

Blicktee replied:
✅

harshavardhana commented on cmd/xl-storage-format-utils.go:
```diff
@@ -52,19 +37,34 @@ func getFileInfoVersions(xlMetaBuf []byte, volume, path string) (FileInfoVersion
+		var latestModTime time.Time
+		if len(versions) > 0 {
+			latestModTime = versions[0].ModTime
+		}
```
```suggestion
		var latestModTime time.Time
		if len(versions) > 0 {
			latestModTime = versions[0].ModTime
		}
```

My suggestion is to return right away like 

```
if len(versions) == 0 {
   return FileInfoVersions{}, nil
}
```

Avoids filling FileInfoVersions() with unexpectedly incomplete data while returning.

klauspost committed:
Return early on 0 versions.

klauspost committed:
Return early on 0 versions.

klauspost committed:
Add InlineData flag

klauspost committed:
Add InlineData flag

klauspost committed:
Merge branch 'master' into fast-multiversion-metadata

klauspost committed:
Merge branch 'master' into fast-multiversion-metadata

klauspost committed:
Add decoding to xl-meta

klauspost committed:
Add decoding to xl-meta

krisis commented on cmd/xl-storage-format-utils.go:
```diff
@@ -52,24 +35,35 @@ func getFileInfoVersions(xlMetaBuf []byte, volume, path string) (FileInfoVersion
 		}
 	}
 	fivs.Versions = fivs.Versions[:n]
+	// Update numversions
```
thank you for fixing this bug  :+1:

krisis commented on cmd/xl-storage-format-v2.go:
```diff
@@ -252,434 +309,428 @@ func (j xlMetaV2Version) getModTime() time.Time {
-func (x xlMetaInlineData) versionOK() bool {
-	if len(x) == 0 {
-		return true
+// getModTime will return the ModTime of the underlying version.
```
```suggestion
// getVersionID will return the versionID of the underlying version.
```

harshavardhana committed:
Merge branch 'master' into fast-multiversion-metadata

klauspost committed:
Add 4 byte signature

klauspost committed:
Add 4 byte signature

klauspost committed:
Merge branch 'fast-multiversion-metadata' of github.com:klauspost/minio into fast-multiversion-metadata

klauspost committed:
Chang receiver names.

klauspost committed:
Chang receiver names.

harshavardhana commented on cmd/xl-storage-format-v2.go:
```diff
@@ -585,29 +664,29 @@ func readXLMetaNoData(r io.Reader, size int64) ([]byte, error) {
 
-func decodeXlHeaders(buf []byte) (versions int, b []byte, err error) {
-	hdrVer, buf, err := msgp.ReadUintBytes(buf)
+func decodeXlHeaders(buf []byte) (versions int, headerV, metaV uint8, b []byte, err error) {
```
```suggestion
func decodeXLHeaders (buf []byte) (versions int, headerV, metaV uint8, b []byte, err error) {
```

klauspost committed:
Bump known xlHeaderVersion

klauspost committed:
Bump known xlHeaderVersion

klauspost committed:
Replace map[string]... hashing.

Before/After:

```
Benchmark_xlMetaV2Shallow_Load/legacy-32         	     468	   2518826 ns/op	   0.34 MB/s	 1326181 B/op	   21391 allocs/op
Benchmark_xlMetaV2Shallow_Load/legacy-32         	     976	   1198166 ns/op	   0.71 MB/s	  377842 B/op	    7709 allocs/op
```

klauspost committed:
Replace map[string]... hashing.

Before/After:

```
Benchmark_xlMetaV2Shallow_Load/legacy-32         	     468	   2518826 ns/op	   0.34 MB/s	 1326181 B/op	   21391 allocs/op
Benchmark_xlMetaV2Shallow_Load/legacy-32         	     976	   1198166 ns/op	   0.71 MB/s	  377842 B/op	    7709 allocs/op
```

klauspost committed:
Update xxh3.

klauspost committed:
Update xxh3.

harshavardhana committed:
Merge branch 'master' into fast-multiversion-metadata

harshavardhana committed:
Update cmd/xl-storage-format-v2.go

harshavardhana committed:
Update cmd/xl-storage-format-v2.go

klauspost committed:
Fix operator order

klauspost committed:
Fix operator order

klauspost committed:
Merge branch 'fast-multiversion-metadata' of github.com:klauspost/minio into fast-multiversion-metadata

klauspost committed:
Add hashDeterministicString unit test.

klauspost committed:
Add hashDeterministicString unit test.

klauspost committed:
Make key+value non-swappable. Pretty print headers.

klauspost committed:
Make key+value non-swappable. Pretty print headers.

klauspost committed:
Switch to fixed size CRC.

klauspost committed:
Switch to fixed size CRC.

klauspost committed:
Faster signatures:

```
Benchmark_xlMetaV2Shallow_Load/legacy-32         	     960	   1218031 ns/op	   0.70 MB/s	  377858 B/op	    7709 allocs/op
Benchmark_xlMetaV2Shallow_Load/legacy-32         	    1238	    950325 ns/op	   0.90 MB/s	  115829 B/op	    5999 allocs/op
```

klauspost committed:
Faster signatures:

```
Benchmark_xlMetaV2Shallow_Load/legacy-32         	     960	   1218031 ns/op	   0.70 MB/s	  377858 B/op	    7709 allocs/op
Benchmark_xlMetaV2Shallow_Load/legacy-32         	    1238	    950325 ns/op	   0.90 MB/s	  115829 B/op	    5999 allocs/op
```

harshavardhana committed:
fix: Makefile conflict

harshavardhana committed:
fix: gocritic cleanups

klauspost committed:
Merge branch 'fast-multiversion-metadata' of github.com:klauspost/minio into fast-multiversion-metadata

# Conflicts:
#	Makefile
#	cmd/xl-storage-format-utils.go
#	cmd/xl-storage-format-v1.go
#	cmd/xl-storage-format-v2.go

klauspost committed:
Merge branch 'master' into fast-multiversion-metadata

klauspost committed:
Remove duplicate after merge.

harshavardhana committed:
Merge branch 'master' into fast-multiversion-metadata

harshavardhana committed:
remove sort

harshavardhana reviewed (approved):
LGTM tested version upgrades from 2019 to this PR works fine. 

Tested `mc admin heal -r` as well to trigger the XLV1 -> XLV2 verified the content is intact and readable.

poornas commented on cmd/erasure-metadata.go:
```diff
@@ -476,9 +476,7 @@ func GetInternalReplicationState(m map[string][]byte) ReplicationState {
 // getInternalReplicationState fetches internal replication state from the map m
 func getInternalReplicationState(m map[string]string) ReplicationState {
-	d := ReplicationState{
-		ResetStatusesMap: make(map[string]string),
```
ResetStatusesMap is used without checking if map is non nil since it is expected to be allocated.Changes needed in bucket-replication-utils.go wherever referenced

harshavardhana replied:
Isn't line #499 expected to do that @poornas

poornas replied:
on closer look https://github.com/minio/minio/blob/master/cmd/bucket-replication-utils.go#L383 uses ResetStatusesMap that is allocated, so non-issue here.

krisis reviewed (approved):
Great work! Tested this PR specifically with ILM transition feature and objects from previous versions of MinIO. Looks good to me :+1:

klauspost committed:
Update cmd/xl-storage-format-v2.go

Co-authored-by: Krishnan Parthasarathi <krisis@users.noreply.github.com>

poornas reviewed (approved):
LGTM
